code/iaas/simple-allocator/src/main/java/io/cattle/platform/simple/allocator/SimpleAllocator.java

```diff
@@ -435,4 +450,76 @@ private ResourceRequest populateResourceRequestFromInstance(Instance instance, S
     }
+
+    @Override
+    protected boolean useLegacyPortAllocation(Long accountId, List<Instance> instances) {
+        Long agentId = getAgentResource(accountId, instances);
+        return useLegacyPortAllocation(agentId);
+    }
+
+    protected boolean useLegacyPortAllocation(Long agentId) {
+        if (agentId == null) {
+            // No scheduler, use legacy logic
+            return true;
+        }
+
+        String schedulerVersion = getSchedulerVersion(agentId);
+        return useLegacyPortAllocation(schedulerVersion);
+    }
+
+    protected String getSchedulerVersion(Long agentId) {
+        Instance instance = agentInstanceDao.getInstanceByAgent(agentId);
+        String imageUuid = (String) DataAccessor.fields(instance).withKey(InstanceConstants.FIELD_IMAGE_UUID).get();
+        DockerImage img = DockerImage.parse(imageUuid);
+        String[] imageParts = img.getFullName().split(":");
+        if (imageParts.length <= 1) {
+            return "";
+        }
+        return imageParts[imageParts.length - 1];
+    }
+
+    protected boolean useLegacyPortAllocation(String actualVersion) {
+        String requiredVersion = PORT_SCHEDULER_IMAGE_VERSION.get();
+        if (StringUtils.isEmpty(requiredVersion)) {
+            // Property not available. Use legacy
+            return true;
+        }
+        String[] requiredParts = requiredVersion.split("\\.");
+        if (requiredParts.length < 3) {
+            // Required image is not following semantic versioning. Assume custom, don't use legacy
+            return false;
+        }
+        int requiredMajor, requiredMinor = 0;
+        try {
+            String majorTemp = requiredParts[0].startsWith("v") ? requiredParts[0].substring(1, requiredParts[0].length()) : requiredParts[0];
+            requiredMajor = Integer.valueOf(majorTemp);
+            requiredMinor = Integer.valueOf(requiredParts[1]);
+        } catch (NumberFormatException e) {
+            // Require image is not following semantic versioning. Assume custom, don't use legacy
+            return false;
+        }
+
+        String[] actualParts = actualVersion.split("\\.");
+        if (actualParts.length < 3) {
+            // Image is not following semantic versioning. Assume custom, don't use legacy
+            return false;
+        }
+
+        int actualMajor, actualMinor = 0;
+        try {
+            String majorTemp = actualParts[0].startsWith("v") ? actualParts[0].substring(1, actualParts[0].length()) : actualParts[0];
+            actualMajor = Integer.valueOf(majorTemp).intValue();
+            actualMinor = Integer.valueOf(actualParts[1]).intValue();
+        } catch (NumberFormatException e) {
+            // Image is not following semantic versioning. Assume custom, don't use legacy
+            return false;
+        }
+
+        if (actualMajor < requiredMajor) {
+            return true;
+        } else if (actualMajor == requiredMajor && actualMinor < requiredMinor) {
```

Comment:
we should perhaps use .intValue() when do Integer comparison. Or Integer1.equals(Integer2)